### PR TITLE
security: remediate findings from Esler architecture review

### PR DESCRIPTION
**Pull request by jamestexas**
## Summary

Addresses all 13 findings from [Mark Esler's architecture review](signet-review.md) dated 2026-04-07. Three findings were already fixed; this PR resolves the remaining 10 with code changes and strengthened documentation.

**Net: -258 lines** (deleted 297 lines of dead code/scaffolding, added 120 lines of fixes)

### HIGH severity fixes
- **#1** Fix Redis token ID — use full JTI (`hex(JTI)` = 32 chars) instead of truncated `EphemeralKeyID[:8]` (16 chars). Redis store was broken — tokens stored via Redis could never be retrieved by middleware.
- **#3** Delete dead `NewProofVerifier` stub (`epr/verifier.go`) that returns `nil` — any caller would crash. Real verifier is in `proof.go`.
- **#4** Fix `CompareSignatures` timing side-channel — swap `bytes.Equal` for `subtle.ConstantTimeCompare`.
- **#14** Strengthen ML-DSA-44 zeroization documentation — the `Pack()` pattern zeros a copy, not the key. Documented impact levels and mitigations.

### MED severity fixes
- **#5** Enforce `SkipPaths` and `RequiredPurposes` in `ServeHTTP` — config accepted these but handler never checked them.
- **#6** Delete unused `settings.go` scaffolding (217 lines of aspirational code with placeholder `Validate()` and `ApplyToConfig()`).
- **#7** Fix zeroize dead-store elimination — replace `runtime.KeepAlive` pattern with `unsafe.Pointer` volatile writes (same technique as Go stdlib `crypto/internal/fips140`).
- **#8** Add `MaxTokenLifetime` (10 min) validation — reject tokens with validity > 10 minutes. Prevents malicious issuers creating year-long tokens.
- **#9** Fix nonce same-second collision — include signature hash in nonce key (`tokenID:timestamp:sigHash`).

### LOW severity fixes
- **#11** Add 4-byte length prefix to purpose field in binding message — prevents boundary ambiguity under extension.
- **#12** Remove fragile `containsNotFound()` string matching in revocation checker — rely on sentinel errors only.
- **#16** Document CBOR canonical encoding as forward-compat constraint on bundle signatures.

### Already fixed (verified in this PR)
- #2 (capability ID separators), #10 (goroutine leak — `Close()` exists), #13 (halfL constant test exists), #15 (IsRevoked fails-closed)

## Breaking changes
- Binding message format changed (length-prefixed purpose). Existing binding signatures will not verify. Acceptable for alpha.
- Tokens with validity > 10 minutes now rejected. Test fixtures updated.

## Test plan
- [x] `go test ./...` — all packages pass
- [x] `go build ./cmd/signet` — binary builds
- [ ] CI passes
- [ ] Integration tests in Docker (`task integration-test`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/http/middleware/signet.go [hunk 1]
```diff
@@ -80,6 +82,16 @@ type signetHandler struct {
...
 // ServeHTTP implements http.Handler with full two-step verification
 func (h *signetHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if this path should skip authentication
+	if len(h.config.SkipPaths) > 0 {
+		for _, path := range h.config.SkipPaths {
+			if r.URL.Path == path || strings.HasPrefix(r.URL.Path, path) {
+				h.next.ServeHTTP(w, r)
+				return
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jamestexas) [reply pkg/http/middleware/signet.go:103]: Fixed in 20f7829 — implemented path-segment-aware matching via `matchesSkipPath()`. Rejects empty and non-slash-prefixed entries. Requires exact match or prefix followed by `/` (so `/api` matches `/api/foo` but not `/api-private`). Added 7 sub-tests covering all edge cases.

File: pkg/http/middleware/signet.go [hunk 2]
```diff
@@ -264,6 +278,24 @@ func (h *signetHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+				break
+			}
+		}
+		if !purposeAllowed {
+			h.config.Logger.Warn("token purpose not allowed", "token_id", tokenID, "purpose", record.Purpose, "required", h.config.RequiredPurposes)
+			h.config.Observer.OnAuthFailure(ctx, ErrPurposeMismatch, "purpose_check")
+			h.config.Metrics.RecordAuthResult("purpose_mismatch", time.Since(startTime))
+			h.config.ErrorHandler(w, r, ErrPurposeMismatch)
+			return
+		}
```
[the remark this answers is not in the document]
Comment (jamestexas) [reply pkg/http/middleware/signet.go:307]: Fixed in 20f7829 — added `TestSignetMiddleware_RequiredPurposes_Allowed` (matching purpose, expects 200) and `TestSignetMiddleware_RequiredPurposes_Rejected` (non-matching purpose, expects 403).

File: pkg/signet/token.go [hunk 3]
```diff
@@ -170,6 +175,8 @@ func (t *Token) validate() error {
 		return fmt.Errorf("%w: issued-at precedes not-before", ErrInvalidToken)
 	case t.ExpiresAt < t.IssuedAt:
 		return fmt.Errorf("%w: expires-at precedes issued-at", ErrInvalidToken)
+	case t.ExpiresAt-t.NotBefore > int64(MaxTokenLifetime.Seconds()):
```
[the remark this answers is not in the document]
Comment (jamestexas) [reply pkg/signet/token.go]: Fixed in 20f7829 — switched to `int64(MaxTokenLifetime/time.Second)` which uses integer division. No float truncation.

File: pkg/http/middleware/redis.go [hunk 4]
```diff
@@ -201,9 +201,10 @@ type storedTokenRecord struct {
 	Metadata                    map[string]string `json:"metadata,omitempty"`
 }
 
-// generateTokenID generates a consistent token ID from a record
+// generateTokenID generates a consistent token ID from a record.
+// Uses the full 16-byte JTI (32 hex chars) to match MemoryTokenStore and middleware lookup.
 func generateTokenID(record *TokenRecord) string {
-	return hex.EncodeToString(record.Token.EphemeralKeyID[:min(8, len(record.Token.EphemeralKeyID))])
+	return hex.EncodeToString(record.Token.JTI)
 }
```
[the remark this answers is not in the document]
Comment (jamestexas) [reply pkg/http/middleware/redis.go:213]: Fixed in 20f7829 — added nil/empty JTI validation in `Store()` before calling `generateTokenID()`, matching the MemoryTokenStore guard.
